plan/statscache/statscache.go

```diff
@@ -14,89 +14,109 @@
 package statscache
 
 import (
+	"fmt"
 	"sync"
-	"time"
 
+	"github.com/juju/errors"
 	"github.com/ngaut/log"
 	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/infoschema"
 	"github.com/pingcap/tidb/meta"
 	"github.com/pingcap/tidb/model"
 	"github.com/pingcap/tidb/plan/statistics"
-	"github.com/pingcap/tidb/store/tikv/oracle"
+	"github.com/pingcap/tidb/util/sqlexec"
 )
 
 type statsInfo struct {
-	tbl      *statistics.Table
-	loadTime int64
+	tbl     *statistics.Table
+	version uint64
 }
 
-type statsCache struct {
-	m     sync.RWMutex
-	cache map[int64]statsInfo
+// Handle can update stats info periodically.
+type Handle struct {
+	ctx         context.Context
+	lastVersion uint64
 }
 
-var statsTblCache = statsCache{cache: map[int64]statsInfo{}}
-
-// expireDuration is 1 hour.
-var expireDuration int64 = 60 * 60 * 1000
-
-func tableCacheExpired(si statsInfo) bool {
-	duration := oracle.GetPhysical(time.Now()) - si.loadTime
-	if duration >= expireDuration {
-		return true
-	}
-	return false
+// NewHandle creates a Handle for update stats.
+func NewHandle(ctx context.Context) *Handle {
+	return &Handle{ctx: ctx}
 }
 
-// GetStatisticsTableCache retrieves the statistics table from cache, and will refill cache if necessary.
-func GetStatisticsTableCache(ctx context.Context, tblInfo *model.TableInfo) *statistics.Table {
-	statsTblCache.m.RLock()
-	si, ok := statsTblCache.cache[tblInfo.ID]
-	statsTblCache.m.RUnlock()
-	// Here we check the TableInfo because there may be some ddl changes in the duration period.
-	// Also, we rely on the fact that TableInfo will not be same if and only if there are ddl changes.
-	if ok && tblInfo == si.tbl.Info && !tableCacheExpired(si) {
-		return si.tbl
+// Update reads stats meta from store and updates the stats map.
+func (h *Handle) Update(m *meta.Meta, is infoschema.InfoSchema) error {
+	sql := fmt.Sprintf("SELECT version, table_id from mysql.stats_meta where version > %d order by version", h.lastVersion)
```

Comment:
stats info version will increase when the table is analyzed. The version is `analyze command`'s startTS.